Project-Hotpot/Project-Hotpot/Managers/CacheManager.swift

```diff
@@ -0,0 +1,128 @@
+                NSLog("Cache Hit")
+                completion(.success(songDetailsArray))
+            }
+        }
+    }
+    func removeSongFromCache(genre: String, song: SongDetails) {
+        cache.evictFromCache(genre: genre, evictSong: song)
+    }
+    func refillCache(genre:String, completion: @escaping (_ result: Result<[SongDetails], Error>) -> Void) {
+        fetchNSongs(limit:SONGS_PER_GENRE, genre:genre) { result in
+            switch result {
+            case .failure(let error):
+                completion(.failure(error))
+            case .success(let dictionary):
+                guard let tracks = dictionary["tracks"] as? [String:Any]?,
+                      let items = tracks?["items"] as? [[String:Any]?]
+                else{
+                    NSLog("failed parsing random song dictionary response")
+                    return completion(.failure(CustomError.failedResponseParsing))
+                }
+                let trackIDs : [String] = items.compactMap{ item in item?["id"] as? String }
+                self.spotifyIdToSongDetails(ids: trackIDs) { result in
+                    switch result{
+                    case .success(let songDetailsArray):
+                        self.cache.addToCache(genre: genre, songs: songDetailsArray)
+                        completion(.success(songDetailsArray))
+                    case .failure(let error) :
+                        completion(.failure(error))
+                    }
+                }
+            }
+        }
+    }
+    
+    func spotifyIdToSongDetails(ids: [String], completion: @escaping (_ result: Result<[SongDetails], Error>) -> Void) {
+        SpotifyManager.shared().fetchAudioFeaturesFromTracks(for: ids) { result in
+            switch result{
+            case .success(let dictionary):
+                guard let features = dictionary["audio_features"] as? [[String:Any]]
+                else {
+                    NSLog("failed to extract features dictionary")
+                    return
+                    
+                }
+                let songDetailsArray:[SongDetails]? = features.compactMap { feature in
+                    guard let featureDictionary = feature as? [String: Any],
+                          let uri = featureDictionary["uri"] as? String,
+                          let id = featureDictionary["id"] as? String,
+                          let danceability = featureDictionary["danceability"] as? NSNumber,
+                          let energy = featureDictionary["energy"] as? NSNumber,
+                          let tempo = featureDictionary["tempo"] as? NSNumber,
+                          let key = featureDictionary["key"] as? NSNumber
+                    else {
+                        NSLog("failed parse of audio feature dictionary")
+                        return nil
+                    }
+                    return SongDetails(uri: uri, id: id, danceability: danceability, energy: energy, tempo: tempo, key: key)
```

Comment:
`Codable` should allow you to inflate an instance of `SongDetails` directly from a `feature`.

This article is biased towards JSON, but is a good overview of `Codable` and what it looks like: https://www.hackingwithswift.com/articles/119/codable-cheat-sheet